File: citizen/citizen.go
package citizen

import (
	"economy/citizen/consumer"
	"economy/citizen/demographics"
	"economy/citizen/sentiment"
	"economy/citizen/taxation"
)

// CitizenReport captures the public economic telemetry of the citizen population.
type CitizenReport struct {
	Population            float64            // Total headcount (e.g. 340M)
	LaborForce            float64            // Civilian labor force
	EmployedCount         float64            // Working headcount
	UnemploymentRate      float64            // U3 rate
	AverageHourlyEarnings float64            // Average hourly wage (USD)
	DisposableIncome      float64            // Aggregate after-tax income (USD)
	ConsumerConfidence    float64            // 0-100 scale
	Happiness             float64            // 0-100 scale
	ConsumerSpending      float64            // Annualized PCE spending (USD)
	TaxesPaidThisTick     float64            // Federal/State taxes paid this tick
	SectorDemand          map[string]float64 // Sector -> demand multiplier
}

// CitizenEconomy coordinates all citizen demographic, psychological, and fiscal behavior.
type CitizenEconomy struct {
	Demographics *demographics.Demographics
	Sentiment    *sentiment.Sentiment
	Taxation     *taxation.Taxation
	Consumer     *consumer.ConsumerEngine
}

// NewDefaultCitizenEconomy constructs an American-style citizen ecosystem.
func NewDefaultCitizenEconomy() *CitizenEconomy {
	return &CitizenEconomy{
		Demographics: demographics.NewDefaultDemographics(),
		Sentiment:    sentiment.NewDefaultSentiment(),
		Taxation:     taxation.NewDefaultTaxation(),
		Consumer:     consumer.NewDefaultConsumerEngine(),
	}
}

// Tick advances the citizen population, sentiment, taxes, and consumer spending.
func (c *CitizenEconomy) Tick(employmentCount, wageRate, inflationRate, wageGrowth, marketReturn, tickFractionOfYear float64) CitizenReport {
	// 1. Advance demographics & calculate disposable income
	effectiveTax := c.Taxation.EffectiveIncomeTaxRate
	c.Demographics.Tick(employmentCount, wageRate, effectiveTax, tickFractionOfYear)

	// 2. Advance sentiment / happiness
	unemp := c.Demographics.UnemploymentRate()
	c.Sentiment.Tick(unemp, inflationRate, wageGrowth, marketReturn, effectiveTax)

	// 3. Compute consumer demand across sectors
	c.Consumer.Tick(
		c.Demographics.AggregateDisposableIncome,
		c.Sentiment.SpendingPropensity,
		c.Demographics.PersonalSavingsRate,
	)

	// 4. Compute taxes paid on earnings and consumption
	grossIncome := c.Demographics.EmployedCount * c.Demographics.AverageHourlyEarnings * 2000.0 * tickFractionOfYear
	incomeTax := c.Taxation.ComputeIncomeTax(grossIncome)
	salesTax := c.Taxation.ComputeSalesTax(c.Consumer.TotalConsumerSpending * tickFractionOfYear)
	c.Taxation.RecordTaxes(incomeTax, 0, salesTax)

	// Clone sector demand map for the report
	sectorDemand := make(map[string]float64, len(c.Consumer.SectorDemandFactors))
	for k, v := range c.Consumer.SectorDemandFactors {
		sectorDemand[k] = v
	}

	return CitizenReport{
		Population:            c.Demographics.TotalPopulation,
		LaborForce:            c.Demographics.LaborForce(),
		EmployedCount:         c.Demographics.EmployedCount,
		UnemploymentRate:      unemp,
		AverageHourlyEarnings: c.Demographics.AverageHourlyEarnings,
		DisposableIncome:      c.Demographics.AggregateDisposableIncome,
		ConsumerConfidence:    c.Sentiment.Index,
		Happiness:             c.Sentiment.Happiness,
		ConsumerSpending:      c.Consumer.TotalConsumerSpending,
		TaxesPaidThisTick:     incomeTax + salesTax,
		SectorDemand:          sectorDemand,
	}
}

// SetPublicFactors routes national health and education levels to citizen demographics.
func (c *CitizenEconomy) SetPublicFactors(healthcareLevel, educationLevel float64) {
	if c.Demographics != nil {
		c.Demographics.SetPublicFactors(healthcareLevel, educationLevel)
	}
}

File: src/citizen/citizen.rs
use std::collections::HashMap;
use serde::{Deserialize, Serialize};

use super::consumer::ConsumerEngine;
use super::demographics::Demographics;
use super::sentiment::Sentiment;
use super::taxation::Taxation;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CitizenReport {
    pub population: f64,
    pub labor_force: f64,
    pub employed_count: f64,
    pub unemployment_rate: f64,
    pub average_hourly_earnings: f64,
    pub disposable_income: f64,
    pub consumer_confidence: f64,
    pub happiness: f64,
    pub consumer_spending: f64,
    pub taxes_paid_this_tick: f64,
    pub sector_demand: HashMap<String, f64>,
}

#[derive(Debug, Clone)]
pub struct CitizenEconomy {
    pub demographics: Demographics,
    pub sentiment: Sentiment,
    pub taxation: Taxation,
    pub consumer: ConsumerEngine,
}

impl Default for CitizenEconomy {
    fn default() -> Self {
        Self {
            demographics: Demographics::default(),
            sentiment: Sentiment::default(),
            taxation: Taxation::default(),
            consumer: ConsumerEngine::default(),
        }
    }
}

impl CitizenEconomy {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn tick(
        &mut self,
        employment_count: f64,
        wage_rate: f64,
        inflation_rate: f64,
        wage_growth: f64,
        market_return: f64,
        tick_fraction_of_year: f64,
    ) -> CitizenReport {
        let effective_tax = self.taxation.effective_income_tax_rate;
        self.demographics.tick(employment_count, wage_rate, effective_tax, tick_fraction_of_year);

        let unemp = self.demographics.unemployment_rate();
        self.sentiment.tick(unemp, inflation_rate, wage_growth, market_return, effective_tax);

        self.consumer.tick(
            self.demographics.aggregate_disposable_income,
            self.sentiment.spending_propensity,
            self.demographics.personal_savings_rate,
        );

        let gross_income = self.demographics.employed_count
            * self.demographics.average_hourly_earnings
            * 2000.0
            * tick_fraction_of_year;
        let income_tax = self.taxation.compute_income_tax(gross_income);
        let sales_tax = self.taxation.compute_sales_tax(self.consumer.total_consumer_spending * tick_fraction_of_year);
        self.taxation.record_taxes(income_tax, 0.0, sales_tax);

        CitizenReport {
            population: self.demographics.total_population,
            labor_force: self.demographics.labor_force(),
            employed_count: self.demographics.employed_count,
            unemployment_rate: unemp,
            average_hourly_earnings: self.demographics.average_hourly_earnings,
            disposable_income: self.demographics.aggregate_disposable_income,
            consumer_confidence: self.sentiment.index,
            happiness: self.sentiment.happiness,
            consumer_spending: self.consumer.total_consumer_spending,
            taxes_paid_this_tick: income_tax + sales_tax,
            sector_demand: self.consumer.sector_demand_factors.clone(),
        }
    }

    pub fn set_public_factors(&mut self, healthcare_level: f64, education_level: f64) {
        self.demographics.set_public_factors(healthcare_level, education_level);
    }
}
